### PR TITLE
AMQP-579: @SendTo - Support Runtime SpEL

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-579

`@SendTo` currently supports initialization time SpEL (`#{...}`).

Add support for runtime SpEL evaluation against the request and reply objects.

This change also eliminates the `ThreadLocal` for `@SendTo`s in the
`DelegatingInvocableHandler` for multi-method listeners, in favor of
returning a `ResultHolder` containg the result and sendTo (when there
is no `replyTo` property.

### CODE REVIEW

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/AbstractAdaptableMessageListener.java [hunk 1]
```diff
@@ -104,6 +123,28 @@ public void setResponseExchange(String responseExchange) {
...
+	 * that return result objects, which will be wrapped in
+	 * a response message and sent to a response destination.
+	 * <p>
+	 * Can be a string starting with "SpEL:" in which case the expression is
+	 * evaluated at runtime; see the reference manual for more information.
+	 * @param defaultReplyTo The exchange.
+	 * @since 1.6
+	 */
+	public void setResponseAddress(String defaultReplyTo) {
+		if (defaultReplyTo.startsWith("SpEL:")) {
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/AbstractAdaptableMessageListener.java]: That smells a bit.
How about to use:
```
PARSER.parseExpression(defaultReplyTo, new TemplateParserContext());
```
instead?
And rely on the standard SpEL syntax: `#{myBean.foo}` etc.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/DelegatingInvocableHandler.java [hunk 2]
```diff
@@ -125,22 +125,32 @@ protected InvocableHandlerMethod getHandlerForPayload(Class<? extends Object> pa
...
-				}
-				Address replyTo = destinations.length == 1 ? new Address(resolve(destinations[0])) : null;
-				if (replyTo != null) {
-					this.defaultReplyToForHandler.put(handler, replyTo);
-				}
-			}
+			replyTo = extractSendTo(method.toString(), ann);
+		}
+		if (replyTo == null) {
+			SendTo ann = AnnotationUtils.getAnnotation(this.bean.getClass(), SendTo.class);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/DelegatingInvocableHandler.java]: `@SendTo` is only method level annotation:
```
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
@Documented
public @interface SendTo {
```
Or have you raised an SF JIRA on the matter already ?

Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/DelegatingInvocableHandler.java]: Oh! I really have a bad memory...
https://jira.spring.io/browse/SPR-13578.
Never mind.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/AbstractAdaptableMessageListener.java (hunk 1 above)
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/AbstractAdaptableMessageListener.java]: OK. I see your point around `resolve()` in the `MethodRabbitListenerEndpoint`, which with the `StandardBeanExpressionResolver` and its internal `beanExpressionParserContext` eats my case.
Anyway I guess we still can get a gain of the `TemplateParserContext` with the `@{` , `}` placeholder for this kind of expressions. 
At least that is what SpEL recommends to do for custom templates.
WDYT?

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/AbstractAdaptableMessageListener.java]: Yes, makes sense; but I'm not keen on `@`, I'll look for something else.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/DelegatingInvocableHandler.java [hunk 3]
```diff
@@ -125,22 +125,32 @@ protected InvocableHandlerMethod getHandlerForPayload(Class<? extends Object> pa
...
+			replyTo = extractSendTo(this.getBean().getClass().getSimpleName(), ann);
 		}
+		if (replyTo != null) {
+			this.handlerSendTo.put(handler, replyTo);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/DelegatingInvocableHandler.java]: I'd say that parse into an `Expression` would be better here. More logical, than in the `AbstractAdaptableMessageListener`.

**Comment by garyrussell** [comment]
Nice - much cleaner and the template context automatically gives us a literal expression if the prefix is not found.

Pushed

File: src/reference/asciidoc/amqp.adoc [hunk 4]
```diff
@@ -1703,8 +1704,38 @@ public String spelReplyTo() {
...
+[source, java]
+----
+@RabbitListener(queues = "test.sendTo.spel")
+@SendTo("!{'some.reply.queue.with.' + result.queueName}")
+public Bar capitalizeWithSendToSpel(Foo foo) {
+    return processTheFooAndReturnABar(foo);
+}
+----
+
+The runtime nature of the SpEL expression is indicated with the `!{...}` prefix.
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: This isn't `prefix` anymore.
Maybe `delimiters` like it is in the Core SpEL doc:
> Each evaluation block is delimited with prefix and suffix characters that you can define, a common choice is to use `#{ }` as the delimiters

Let me know your thought and I'll fix on merge.

Comment (garyrussell) [inline src/reference/asciidoc/amqp.adoc]: Thanks yes, should be

    is indicated with `!{...}` delimiters.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/30cabc0ac64916cd5e2388f67f72e9a47fac7bcd

**Pull request closed without merging**